File: src/main/java/me/nepnep/nepbot/message/command/commands/Help.java
package me.nepnep.nepbot.message.command.commands;

import me.nepnep.nepbot.message.command.Category;
import me.nepnep.nepbot.message.command.CommandRegister;
import me.nepnep.nepbot.message.command.ICommand;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.TextChannel;
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent;

import java.util.Arrays;
import java.util.List;

public class Help implements ICommand {
    @Override
    public void execute(List<String> args, GuildMessageReceivedEvent event) {
        EmbedBuilder builder = new EmbedBuilder();
        String categories = Arrays.toString(Arrays.stream(Category.values()).map(Enum::toString).toArray(String[]::new));
        TextChannel channel = event.getChannel();

        if (args.size() == 0) {
            String categoriesFormatted = String.format(categories
                    .replace("[", "")
                    .replace("]", "")
                    .replace(", ", "%n"));

            String general = getDescriptions(Category.GENERAL);

            builder.setDescription(String.format("General:%nCategories:%n%n%s%n%nCommands:%n%n%s", categoriesFormatted, general));

            channel.sendMessage(builder.build()).queue();
        } else {
            try {
                int page = Integer.parseInt(args.get(0));
                switch (page) {
                    case 1:
                        builder.setDescription(getDescriptions(Category.ADMIN));
                        channel.sendMessage(builder.build()).queue();
                        break;
                    case 2:
                        builder.setDescription(getDescriptions(Category.ANIMALS));
                        channel.sendMessage(builder.build()).queue();
                        break;
                    case 3:
                        builder.setDescription(String.format("JoinMessage%n%nDefaultRole%n%nLeaveMessage%n%nStarboard"));
                        channel.sendMessage(builder.build()).queue();
                        break;
                    case 4:
                        builder.setDescription(getDescriptions(Category.FUN));
                        channel.sendMessage(builder.build()).queue();
                        break;
                    case 5:
                        builder.setDescription(getDescriptions(Category.INFO));
                        channel.sendMessage(builder.build()).queue();
                        break;
                    case 6:
                        builder.setDescription(String.format("Communism%n%nOh my how lewd%n%nWhat if?"));
                        channel.sendMessage(builder.build()).queue();
                        break;
                    case 7:
                        builder.setDescription(getDescriptions(Category.MOD));
                        channel.sendMessage(builder.build()).queue();
                        break;
                    default:
                        channel.sendMessage("Invalid page!").queue();
                }
            } catch (NumberFormatException e) {
                channel.sendMessage("Invalid page!").queue();
            }
        }
    }
    private String getDescriptions(Category category) {
        StringBuilder builder = new StringBuilder();
        ICommand[] commands = CommandRegister.REGISTER
                .values()
                .stream()
                .filter(command -> command.getCategory().equals(category))
                .toArray(ICommand[]::new);

        for (ICommand command : commands) {
            builder.append(command.getInvoke()).append(String.format("%n%s%n%n", command.getDescription()));
        }
        return builder.toString();
    }
    @Override
    public String getInvoke() {
        return "help";
    }
    @Override
    public Permission requiredPermission() {
        return null;
    }
    @Override
    public Category getCategory() {
        return Category.GENERAL;
    }
    @Override
    public String getDescription() {
        return "Help page: ;help <int page> | null";
    }
}

File: src/main/kotlin/me/nepnep/nepbot/message/command/commands/Help.kt
package me.nepnep.nepbot.message.command.commands

import dev.minn.jda.ktx.events.onStringSelect
import dev.minn.jda.ktx.interactions.components.StringSelectMenu
import dev.minn.jda.ktx.interactions.components.option
import dev.minn.jda.ktx.messages.Embed
import dev.minn.jda.ktx.messages.editMessage_
import dev.minn.jda.ktx.messages.send
import kotlinx.coroutines.delay
import me.nepnep.nepbot.message.command.AbstractCommand
import me.nepnep.nepbot.message.command.Category
import me.nepnep.nepbot.message.command.CommandRegister
import net.dv8tion.jda.api.entities.channel.middleman.GuildMessageChannel
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.dv8tion.jda.api.interactions.components.ActionRow
import java.util.*
import kotlin.time.DurationUnit
import kotlin.time.toDuration

class Help : AbstractCommand(
    "help",
    Category.GENERAL,
    "Help page"
) {
    override suspend fun execute(args: List<String>, event: MessageReceivedEvent, channel: GuildMessageChannel) {
        val id = "help:${UUID.randomUUID()}"
        val menu = createMenu(id, Category.GENERAL)
        
        event.channel.send(embeds = listOf(Embed(getDescriptions(Category.GENERAL))), components = listOf(ActionRow.of(menu))).queue()
        
        val handler = event.jda.onStringSelect(id) {
            val category = Category.valueOf(it.values[0])
            it.editMessage_(
                embeds = listOf(Embed(getDescriptions(category))),
                components = listOf(ActionRow.of(createMenu(id, category)))
            ).queue()
        }
        
        delay(1.toDuration(DurationUnit.MINUTES))
        handler.cancel()
    }

    private fun getDescriptions(category: Category): String {
        if (category == Category.EVENTS) {
            return """
                JoinMessage

                DefaultRole

                LeaveMessage

                Starboard
                
                (Potentially more in plugins)
            """
        }
        
        val commands = CommandRegister.register.values.filter { it.category == category }

        val builder = StringBuilder()
        for (command in commands) {
            builder.append(command.invoke + "\n${command.description}\n\n")
        }
        return builder.toString()
    }
    
    private fun createMenu(id: String, default: Category) = StringSelectMenu(id) {
        Category.values().forEach {
            option(it.toString(), it.name, default = default == it)
        }
    }
}